athena-neptune/src/main/java/com/amazonaws/athena/connectors/neptune/NeptuneRecordHandler.java

```diff
@@ -0,0 +1,248 @@
+     * @param queryStatusChecker A QueryStatusChecker that you can use to stop doing
+     *                           work for a query that has already terminated
+     * @throws IOException
+     * @note Avoid writing >10 rows per-call to BlockSpiller.writeRow(...) because
+     *       this will limit the BlockSpiller's ability to control Block size. The
+     *       resulting increase in Block size may cause failures and reduced
+     *       performance.
+     */
+    @Override
+    protected void readWithConstraint(final BlockSpiller spiller, final ReadRecordsRequest recordsRequest,
+            final QueryStatusChecker queryStatusChecker) {
+        logger.info("readWithConstraint: enter - " + recordsRequest.getSplit());
+        final TableName tableName = recordsRequest.getTableName();
+        final String labelName = tableName.getTableName();
+        long numRows = 0;
+        final AtomicLong numResultRows = new AtomicLong(0);
+        final Client client = neptuneConnection.getNeptunClientConnection();
+
+        try {
+            final GraphTraversalSource graphTraversalSource = neptuneConnection.getTraversalSource(client);
+
+            String traversal = "g.V().hasLabel('" + labelName + "')";
+
+            if (recordsRequest.getConstraints().getSummary().size() > 0) {
+
+                try {
+                    logger.info("readWithContraint: Constaints Map "
+                            + recordsRequest.getConstraints().getSummary().toString());
+
+                    final Map<String, ValueSet> constraints = recordsRequest.getConstraints().getSummary();
+                    traversal += flattenContraintsMap(constraints);
+                } catch (final Exception e) {
+                    logger.info("readWithContraint: Exception occured at typecasting " + e.getStackTrace());
+                }
+            }
+
+            traversal += ".valueMap()";
+
+            logger.info("readWithContraint: Neptune Query " + traversal);
+
+            // Code to build gremlin query from string
+            final ConcurrentBindings b = new ConcurrentBindings();
+            b.putIfAbsent("g", graphTraversalSource);
+
+            final GremlinExecutor ge = GremlinExecutor.build().evaluationTimeout(15000L).globalBindings(b).create();
+
+            final CompletableFuture<Object> evalResult = ge.eval(traversal);
+            final GraphTraversal<Vertex, Map<Object, Object>> result = (GraphTraversal<Vertex, Map<Object, Object>>) evalResult
+                    .get();
+
+            while (result.hasNext() && queryStatusChecker.isQueryRunning()) {
+                numRows++;
+                try {
+                    spiller.writeRows((final Block block, final int rowNum) -> {
+                        final Map<Object, Object> obj = result.next();
+                        boolean matched = true;
+
+                        for (final Field nextField : recordsRequest.getSchema().getFields()) {
```

Comment:
Can we move this to use the RowWriter abstraction? It is much (2x-4x) faster than all the looping and branching being done here: https://github.com/awslabs/aws-athena-query-federation/blob/master/athena-example/src/main/java/com/amazonaws/connectors/athena/example/ExampleRecordHandler.java#L146

Reply:
isn't this hardcoding of field names? 

builder.withExtractor("year", (IntExtractor) (Object context, NullableIntHolder value) -> {
             value.isSet = 1;
             value.value = Integer.parseInt(((String[]) context)[0]);
         });

Reply:
implemented RowWriter pattern

Reply:
nice job.